src/biokbase/narrative/tests/test_appmanager.py

```diff
@@ -117,41 +183,68 @@ def test_run_app_good_inputs(self, c, auth):
         self.assertEqual(new_job.app_id, self.test_app_id)
         self.assertEqual(new_job.tag, self.test_tag)
         self.assertIsNone(new_job.cell_id)
+        self._verify_comm_success(c.return_value.send_comm_message)
 
     @mock.patch("biokbase.narrative.jobs.appmanager.clients.get", get_mock_client)
     @mock.patch("biokbase.narrative.jobs.appmanager.JobComm")
     @mock.patch(
         "biokbase.narrative.jobs.appmanager.auth.get_agent_token",
         side_effect=mock_agent_token,
     )
-    def test_run_app_from_gui_cell(self, c, auth):
-        c.return_value.send_comm_message.return_value = None
-        os.environ["KB_WORKSPACE_ID"] = self.public_ws
+    def test_run_app_from_gui_cell(self, auth, c):
+        cell_id = "12345"
+        c.return_value.send_comm_message = MagicMock()
         self.assertIsNone(
             self.am.run_app(
                 self.test_app_id,
                 self.test_app_params,
                 tag=self.test_tag,
-                cell_id="12345",
+                cell_id=cell_id,
             )
         )
+        self._verify_comm_success(c.return_value.send_comm_message, cell_id=cell_id)
 
     @mock.patch("biokbase.narrative.jobs.appmanager.JobComm")
     def test_run_app_bad_id(self, c):
-        c.return_value.send_comm_message.return_value = None
-        self.assertIsNone(self.am.run_app(self.bad_app_id, None))
+        c.return_value.send_comm_message = MagicMock()
+
+        def run_func():
+            return self.am.run_app(self.bad_app_id, None)
+
+        self.run_app_expect_error(
+            c.return_value.send_comm_message,
+            run_func,
+            "run_app",
+            f'Unknown app id "{self.bad_app_id}" tagged as "release"',
+        )
 
     @mock.patch("biokbase.narrative.jobs.appmanager.JobComm")
     def test_run_app_bad_tag(self, c):
-        c.return_value.send_comm_message.return_value = None
-        self.assertIsNone(self.am.run_app(self.good_app_id, None, tag=self.bad_tag))
+        c.return_value.send_comm_message = MagicMock()
+
+        def run_func():
+            return self.am.run_app(self.good_app_id, None, tag=self.bad_tag)
```

Comment:
You could refactor this even further and just provide the method name and the arguments -- save another couple of precious characters!

Reply:
I'm starting to work on just parameterizing the whole thing. Feed a set of inputs and expected errors and just iterate over them. Will squash 12-15 semi-repetitive tests into ~3, if I get it right.